---
title: "Analysis of Liver Hepatocellular Carcinoma Data in R"
author: 'Bo Yang, Rui Nie, Yueying Hu'
output: pdf_document
---


```{r, message=FALSE, include=FALSE}
rm(list = ls())
# Load packages
library("TCGAbiolinks")
library("limma")
library("edgeR")
library("glmnet")
library("factoextra") # PCA
library("FactoMineR") # PCA
library("caret")
library("SummarizedExperiment")
library("gplots")
library("survival")
library("survminer")
library("RColorBrewer")
library("gProfileR")
library("genefilter")
library("jsonlite")
library("rpart")
library("rpart.plot")
library("randomForest")
library("data.table")
library("clusterProfiler")
library("org.Hs.eg.db") 
# library(countToFPKM)
```


# Introduction

Hepatocellular carcinoma (LIHC/HCC) is the one of the leading cause of cancer death, which accounts for 8.2% of global cancer incidence and 4.7% of cancer-related
mortality []. Due to the high mortality rate, the research on this specific cancer type is significant. Previous study suggested that lifestyles and personal habits, including drinking alcohol, smoking, are positively associated with the onset LIHC [], which has greatly enhanced the efficiency of screening for the disease by health providers. Nevertheless, the difficulty for disease detection remains alarmingly high and current median survival spans around 6 to 20 months. Under such circumstances, an indicator with higher sensitivity would advance the diagnosis stage prolong patients' survival.

The existing treatment options includes local destruction of the tumor, organ replacement, or surgical removal of the tumor. Despite their healing effects and improvement of the quality of life, treatment such as chemotherapy and drug intakes may introduce significant side effects. This reinforces the importance of identifying a predictive biomarker for LIHC. 

RNA Sequencing (RNA-seq) using the capabilities of high-throughput sequencing methods provides insight into the transcriptome of a cell. Through understanding the gene expression patterns reflected by RNA sequencing results, we aim to locate indicator genes for LIHC with the help of machine learning modeling. Furthermore, given the massive amount of existing genes in human genome, our problem can be formulated as a high-dimensional problem. One of our side goal is to perform variable selection to reduce the risk of overfitting.

